lib/luacrypto_skcipher.c

```diff
@@ -0,0 +1,248 @@
+#include <lualib.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+static const lunatik_class_t luacrypto_skcipher_class;
+
+typedef struct {
+	struct crypto_skcipher *tfm;
+	struct skcipher_request *req;
+} luacrypto_skcipher_t;
+
+LUNATIK_PRIVATECHECKER(luacrypto_skcipher_check, luacrypto_skcipher_t *);
+
+static void luacrypto_skcipher_release(void *private)
+{
+	luacrypto_skcipher_t *tfm_ud = (luacrypto_skcipher_t *)private;
+	if (!tfm_ud)
+		return;
+
+	if (tfm_ud->req) {
+		skcipher_request_free(tfm_ud->req);
+		tfm_ud->req = NULL;
+	}
+
+	if (tfm_ud->tfm && !IS_ERR(tfm_ud->tfm))
+		crypto_free_skcipher(tfm_ud->tfm);
+}
+
+/***
+* SKCIPHER Object methods.
+* These methods are available on SKCIPHER TFM objects created by `crypto_skcipher.new()`.
+* @type crypto_skcipher
+*/
+
+/***
+* Sets the encryption key for the SKCIPHER transform.
+* @function crypto_skcipher:setkey
+* @tparam string key The encryption key.
+* @raise Error if setting the key fails (e.g., invalid key length for the algorithm).
+*/
+static int luacrypto_skcipher_setkey(lua_State *L) {
+	luacrypto_skcipher_t *tfm_ud = luacrypto_skcipher_check(L, 1);
+	size_t keylen;
+	const char *key = luaL_checklstring(L, 2, &keylen);
+	lunatik_try(L, crypto_skcipher_setkey, tfm_ud->tfm, key, keylen);
+	return 0;
+}
+
+/***
+* Gets the required initialization vector (IV) size for the SKCIPHER transform.
+* @function crypto_skcipher:ivsize
+* @treturn integer The IV size in bytes.
+*/
+static int luacrypto_skcipher_ivsize(lua_State *L) {
+	luacrypto_skcipher_t *tfm_ud = luacrypto_skcipher_check(L, 1);
+	lua_pushinteger(L, crypto_skcipher_ivsize(tfm_ud->tfm));
+	return 1;
+}
+
+/***
+* Gets the block size of the SKCIPHER transform.
+* Data processed by encrypt/decrypt should typically be a multiple of this size,
+* depending on the cipher mode.
+* @function crypto_skcipher:blocksize
+* @treturn integer The block size in bytes.
+*/
+static int luacrypto_skcipher_blocksize(lua_State *L) {
+	luacrypto_skcipher_t *tfm_ud = luacrypto_skcipher_check(L, 1);
+	lua_pushinteger(L, crypto_skcipher_blocksize(tfm_ud->tfm));
+	return 1;
+}
+
+#define LUACRYPTO_SKCIPHER_CRYPT_FN(name) \
+static int luacrypto_skcipher_##name(lua_State *L) { \
+	luacrypto_skcipher_t *tfm_ud = luacrypto_skcipher_check(L, 1); \
+	size_t iv_s_len, data_s_len; \
+	const char *iv_s = luaL_checklstring(L, 2, &iv_s_len); \
+	const char *data_s = luaL_checklstring(L, 3, &data_s_len); /* Data (plaintext or ciphertext) */ \
+\
+	if (!tfm_ud->req) { \
+		return luaL_error(L, "skcipher_tfm:%s: TFM request not initialized (internal error)", \
+				  #name); \
+	} \
+\
+	unsigned int expected_ivlen = crypto_skcipher_ivsize(tfm_ud->tfm); \
+	luaL_argcheck(L, iv_s_len == expected_ivlen, 2, "incorrect IV length"); \
+\
+	u8 *iv_data = lunatik_checkalloc(L, iv_s_len); \
+	memcpy(iv_data, iv_s, iv_s_len); \
+\
+	luaL_Buffer work_buffer; \
+	struct scatterlist sg_work; \
+\
+	u8 *buf = luaL_buffinitsize(L, &work_buffer, data_s_len); \
```

Comment:
it might raise an error due to mem allocation.. in this case, you will leak iv_data..
https://www.lua.org/manual/5.4/manual.html#luaL_buffinitsize

Reply:
my suggestion is to create an aux inline function to prep request.. you can use try_ret to catch the error and then free iv_data, if it's the case

Reply:
I ended taking this approach: allocate iv_data after buf, just when needed, and freeing it if `crypt_func` fails.